src/main/java/org/apache/spark/shuffle/ucx/rpc/RpcConnectionCallback.java

```diff
@@ -0,0 +1,90 @@
+import org.openucx.jucx.UcxRequest;
+import org.openucx.jucx.ucp.UcpEndpoint;
+import org.openucx.jucx.ucp.UcpEndpointParams;
+import org.openucx.jucx.ucp.UcpWorker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * RPC processing logic. Both driver and excutor accepts the same RPC messgae:
+ * executor worker address followed by it's serialized BlockManagerID.
+ * Executor on accepting this message just adds workerAddress to the connection map.
+ * Driver doing the logic of introducing connected executor to cluster nodes and
+ * introduce cluster to connected executor.
+ */
+public class RpcConnectionCallback extends UcxCallback {
+  private static final Logger logger = LoggerFactory.getLogger(RpcConnectionCallback.class);
+  private final ByteBuffer metadataBuffer;
+  private final boolean isDriver;
+  private final UcxNode ucxNode;
+  private static final ConcurrentMap<UcpEndpoint, ByteBuffer> rpcConnections =
+    UcxNode.getRpcConnections();
+  private static final ConcurrentMap<BlockManagerId, ByteBuffer> workerAdresses =
+    UcxNode.getWorkerAddresses();
+
+  RpcConnectionCallback(ByteBuffer metadataBuffer, boolean isDriver, UcxNode ucxNode) {
+    this.metadataBuffer = metadataBuffer;
+    this.isDriver = isDriver;
+    this.ucxNode = ucxNode;
+  }
+
+  @Override
+  public void onSuccess(UcxRequest request) {
+    int workerAddressSize = metadataBuffer.getInt();
+    ByteBuffer workerAddress = Platform.allocateDirectBuffer(workerAddressSize);
+
+    // Copy worker address from metadata buffer to separate buffer.
+    final ByteBuffer metadataView = metadataBuffer.duplicate();
+    metadataView.limit(metadataView.position() + workerAddressSize);
+    workerAddress.put(metadataView);
+    metadataBuffer.position(metadataBuffer.position() + workerAddressSize);
+
+    BlockManagerId blockManagerId;
+    try {
+      blockManagerId = SerializableBlockManagerID
+        .deserializeBlockManagerID(metadataBuffer);
+    } catch (IOException e) {
+      String errorMsg = String.format("Failed to deserialize BlockManagerId: %s", e.getMessage());
+      throw new UcxException(errorMsg);
+    }
+    logger.debug("Received RPC message from {}", blockManagerId);
+    UcpWorker globalWorker = ucxNode.getGlobalWorker();
+
+    workerAddress.clear();
+
+    if (isDriver) {
+      metadataBuffer.clear();
+      UcpEndpoint newConnection = globalWorker.newEndpoint(
+        new UcpEndpointParams().setPeerErrorHadnlingMode()
+          .setUcpAddress(workerAddress));
+      // For each existing connection
+      rpcConnections.keySet().forEach(connection -> {
+        // send address of joined worker to already connected workers
+        connection.sendTaggedNonBlocking(metadataBuffer, null);
+        // introduce other workers to joined worker
+        newConnection.sendTaggedNonBlocking(rpcConnections.get(connection), null);
```

Comment:
the send operations must be blocking because there is no extra reference for metadataBuffer

Reply:
Metadata buffer is taken from mempool:
```
RegisteredMemory metadataMemory = memoryPool.get(conf.metadataRPCBufferSize());
    ByteBuffer metadataBuffer = metadataMemory.getBuffer();
```

Reply:
when is it recycled to the pool?

Reply:
never mind, i see it, it's ok